### PR TITLE
"Append link" action

### PR DESCRIPTION
**Pull request by bfollington**
Easily append a link to a note you're viewing to another note, useful for "bookmarking" behaviour.

- [x] Code review
- [x] Append text to note BEFORE loading it to ensure it happens only once
- [x] Implement editor and viewer flow
- [x] Implement for all tabs
- [x] Tests (action mapping, append action itself)

https://github.com/subconsciousnetwork/subconscious/assets/5009316/7234109a-3517-4d58-96f3-e91819a81156

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 1]
```diff
@@ -373,7 +374,11 @@ enum MemoEditorDetailAction: Hashable {
     /// more recently than DetailModel.
     case setDetailLastWriteWins(MemoEditorDetailResponse)
     /// Set detail
-    case setDetail(detail: MemoEditorDetailResponse, autofocus: Bool)
+    case setDetail(
+        detail: MemoEditorDetailResponse,
+        autofocus: Bool,
+        append: String = ""
+    )
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:381]: Generic approach to appending to whatever note we loaded.

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift [hunk 2]
```diff
@@ -122,6 +192,35 @@ struct MemoViewerDetailMetaSheetModel: ModelProtocol {
...
+            
+        // Rename
+        case .appendLinkSearch(let action):
+            return MemoViewerDetailMetaSheetAppendLinkSearchCursor.update(
+                state: state,
+                action: action,
+                environment: environment
+            )
+        case .presentAppendLinkSearchSheet(let isPresented):
+            return presentRenameSheet(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift:204]: name

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift [hunk 3]
```diff
@@ -144,6 +243,49 @@ struct MemoViewerDetailMetaSheetModel: ModelProtocol {
         model.author = author
         return Update(state: model)
     }
+    
+    static func presentRenameSheet(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift:247]: name

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 4]
```diff
@@ -953,6 +953,85 @@ final class DatabaseService {
...
+    /// Given
+    /// - An entry link representing the current link for the file
+    /// - An entry link representing the search query
+    /// - And some results for existing entries
+    /// ...Return an array of unique RenameSuggestions,
+    /// sorted for presentation.
+    ///
+    /// We factor out this collation code to make it easier to test.
+    ///
+    /// - Returns an array of RenameSuggestion
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:966]: Rewrite

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 5]
```diff
@@ -953,6 +953,85 @@ final class DatabaseService {
...
+    static func collateAppendLinkSuggestions(
+        current: Slashlink,
+        results: [Slashlink]
+    ) -> [AppendLinkSuggestion] {
+        var suggestions: OrderedDictionary<Slug, AppendLinkSuggestion> = [:]
+        
+        // Then append results from existing entries, potentially overwriting
+        // result for literal query if identical.
+        for result in results {
+            /// If slug changed, this is a move
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:976]: Update comments

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 6]
```diff
@@ -953,6 +953,85 @@ final class DatabaseService {
...
+                    ),
+                    forKey: result.slug
+                )
+            }
+        }
+        return Array(suggestions.values)
+    }
+    
+    /// Given a query and a `current` slug, produce an array of suggestions
+    /// for renaming the note.
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:991]: Update comment

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+                        dismiss()
+                    }
+                }
+            })
+        }
+    }
+}
+
+enum AppendLinkSearchAction: Hashable {
+    /// Set the memo for which this rename sheet is being invoked.
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:60]: Update comment

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+struct AppendLinkSearchModel: ModelProtocol {
+    var subject: Slashlink? = nil
+    var queryField: FormField<String, String> = FormField(value: "", validate: { s in s })
+    /// Suggestions for renaming note.
+    var suggestions: [AppendLinkSuggestion] = []
+    
+    static let logger = Logger(
+        subsystem: Config.default.rdns,
+        category: "RenameSearch"
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:79]: Update name

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                action: .setQuery(state.queryField.value),
+                environment: environment
+            )
+        case .setSuggestions(let suggestions):
+            return setSuggestions(
+                state: state,
+                suggestions: suggestions
+            )
+        case .failSuggestions(let error):
+            return failRenameSuggestions(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:118]: name

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    }
+    
+    /// Set text of slug field
+    static func setQuery(
+        state: Self,
+        environment: AppEnvironment,
+        query: String
+    ) -> Update<Self> {
+        guard let current = state.subject else {
+            logger.log("Rename query updated, but no subject set. Doing nothing.")
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:146]: Update log

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        .eraseToAnyPublisher()
+        
+        return update(
+            state: state,
+            action: .queryField(.setValue(input: query)),
+            environment: environment
+        ).mergeFx(fx)
+    }
+    
+    /// Set rename suggestions
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:175]: Update comment

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+        suggestions: [AppendLinkSuggestion]
+    ) -> Update<Self> {
+        var model = state
+        model.suggestions = suggestions
+        return Update(state: model)
+    }
+
+    /// Handle rename suggestions error.
+    /// This case can happen e.g. if the database fails to respond.
+    static func failRenameSuggestions(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:187]: Update name / comment

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift [hunk 13]
```diff
@@ -76,11 +104,50 @@ enum MemoViewerDetailMetaSheetAction: Hashable {
     case setAuthor(_ author: UserProfile)
     case requestDismiss
     case requestAuthorDetail(_ author: UserProfile)
+    
+    /// Tagged actions for rename search sheet
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift:108]: Update comment

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift [hunk 14]
```diff
@@ -76,11 +104,50 @@ enum MemoViewerDetailMetaSheetAction: Hashable {
...
+    case appendLinkSearch(AppendLinkSearchAction)
+    case presentAppendLinkSearchSheet(_ isPresented: Bool)
+    case presentAppendLinkSearchSheetFor(_ address: Slashlink?)
+    case selectAppendLinkSearchSuggestion(AppendLinkSuggestion)
+    
+    static func setRenameSearchSubject(_ address: Slashlink?) -> Self {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift:114]: Update name

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift [hunk 15]
```diff
@@ -76,11 +104,50 @@ enum MemoViewerDetailMetaSheetAction: Hashable {
...
+        .appendLinkSearch(.setSubject(address))
+    }
+}
+
+// MARK: RenameSearchCursor
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift:119]: Update comment

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift [hunk 16]
```diff
@@ -122,6 +192,35 @@ struct MemoViewerDetailMetaSheetModel: ModelProtocol {
             return Update(state: state)
         case .requestAuthorDetail:
             return Update(state: state)
+            
+        // Rename
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailMetaSheetView.swift:196]: Update name

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 17]
```diff
@@ -3024,6 +3046,27 @@ struct AppModel: ModelProtocol {
...
+            return .succeedAppendToEntry(address: address)
+        }
+        .recover { error in
+            .failAppendToEntry(address: address, error.localizedDescription)
+        }
+        .eraseToAnyPublisher()
+        
+        return Update(state: state, fx: fx)
+    }
+
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/AppView.swift:3069]: Generic machinery to append text to a note, I can see this being something we for more than links.

File: xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift [hunk 18]
```diff
@@ -271,8 +274,31 @@ struct DetailStackModel: Hashable, ModelProtocol {
...
+            return update(
+                state: state,
+                action: .pushDetail(
+                    .editor(
+                        MemoEditorDetailDescription(
+                            address: address
+                        )
+                    )
+                ),
+                environment: environment
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift:297]: I can see an argument for two UX patterns on appending:

1. open the note, showing the newly appended line, with the cursor focused after the new content
2. append the link and close the metasheet, treating it more like a bookmark and preserving the current context

I am personally leaning towards 2 but undecided.

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 19]
```diff
@@ -149,7 +149,7 @@ struct MemoEditorDetailView: View {
             MemoEditorDetailMetaSheetView(
                 store: store.viewStore(
                     get: \.metaSheet,
-                    tag: DetailMetaSheetCursor.tag
+                    tag: MemoEditorDetailMetaSheetCursor.tag
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:152]: Renamed for clarity (the Viewer version is prefixed)

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 20]
```diff
@@ -450,6 +456,7 @@ enum MemoEditorDetailAction: Hashable {
 
     /// Set selected range in editor
     case setEditorSelection(range: NSRange, text: String)
+    case appendEditorText(text: String)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:459]: drop

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 21]
```diff
@@ -740,6 +750,12 @@ struct MemoEditorDetailModel: ModelProtocol {
                 range: range,
                 text: text
             )
+        case let .appendEditorText(text):
+            return appendEditorText(
+                state: state,
+                environment: environment,
+                text: text
+            )
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:758]: drop

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 22]
```diff
@@ -1163,6 +1188,26 @@ struct MemoEditorDetailModel: ModelProtocol {
...
+                    saveState: .unsaved,
+                    modified: Date.now
+                ),
+                .autosave
+            ],
+            environment: environment
+        )
+    }
+    
     /// Insert text in editor at range
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:1211]: drop

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 23]
```diff
@@ -2101,6 +2135,8 @@ struct MemoEditorDetailModel: ModelProtocol {
 struct MemoEditorDetailDescription: Hashable {
     var address: Slashlink?
     var fallback: String = ""
+    
+    var expectedVersion: Cid? = nil
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:2139]: drop

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+                        },
+                        label: {
+                            AppendLinkSuggestionLabelView(suggestion: suggestion)
+                        }
+                    )
+                    .modifier(SuggestionViewModifier())
+                }
+                .listStyle(.plain)
+            }
+            .navigationTitle("Edit link")
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift:46]: Change to "Append link"

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift [hunk 25]
```diff
@@ -1423,24 +1441,13 @@ struct MemoEditorDetailModel: ModelProtocol {
...
-            // If autofocus is false, request blur if needed
-            return update(
-                state: state,
-                actions: [
-                    .setDetailLastWriteWins(detail),
-                    .requestEditorFocus(false)
-                ],
-                environment: environment
-            )
-        }
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:1436]: Intentional removal?

**Review by gordonbrander — CHANGES_REQUESTED** [review]
Just one change requested. Otherwise LGTM.

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift (hunk 25 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:1436]: Not fully intentional, I forgot to restore this code. It was behaving from a UX POV but I've clarified the code to explode the thinking behind it.

File: xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift (hunk 24 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Detail/AppendLinkSearchView.swift]: Fixed!

**Review by gordonbrander — APPROVED** [review]
LGTM

**Pull request merged**